x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/apikey/RestUpdateApiKeyAction.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.rest.action.apikey;
+
+import org.elasticsearch.client.internal.node.NodeClient;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.license.XPackLicenseState;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.RestToXContentListener;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xpack.core.security.action.apikey.UpdateApiKeyAction;
+import org.elasticsearch.xpack.core.security.action.apikey.UpdateApiKeyRequest;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.security.rest.action.SecurityBaseRestHandler;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+import static org.elasticsearch.rest.RestRequest.Method.PUT;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public final class RestUpdateApiKeyAction extends SecurityBaseRestHandler {
+
+    @SuppressWarnings("unchecked")
+    static final ConstructingObjectParser<Payload, Void> PARSER = new ConstructingObjectParser<>(
+        "update_api_key_request_payload",
+        a -> new Payload((List<RoleDescriptor>) a[0], (Map<String, Object>) a[1])
+    );
+
+    static {
+        PARSER.declareNamedObjects(optionalConstructorArg(), (p, c, n) -> {
+            p.nextToken();
+            return RoleDescriptor.parse(n, p, false);
+        }, new ParseField("role_descriptors"));
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> p.map(), new ParseField("metadata"));
+    }
+
+    public RestUpdateApiKeyAction(final Settings settings, final XPackLicenseState licenseState) {
+        super(settings, licenseState);
+    }
+
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(POST, "/_security/api_key/_update/{id}"));
```

Comment:
I think we currently have a bug in our route validation: `PUT /_security/api_key/{id}` is not possible here since it clashes with `POST /_security/api_key/{ids}/_clear_cache` because of the different wildcard path parameter names. Starting the node with the PUT route in place leads to an illegal arg exception.

```
fatal exception while booting Elasticsearchjava.lang.IllegalArgumentException: Trying to use conflicting wildcard names for same path: ids and id
```

I don't think this is correct behavior -- the path suffix should be sufficient to distinguish those two routes. I'll reproduce in a test and open an issue.

In any case, that's why I'm going for `POST /_security/api_key/_update/{id}` here for now. Don't think this is a blocker for this PR -- we can always tweak the route, as long as it's before 8.4 release.

Reply:
It's not a bug. The issue is that we have to use the same placeholder name for different APIs that share the same prefix. So it will work if we have the following instead
```
new Route(POST, "/_security/api_key/{ids}")
```
(note `ids` vs `id`)

However, the new API guideline recommends singular form over plural in URL path. So we should instead change `POST /_security/api_key/{ids}/_clear_cache` to `POST /_security/api_key/{id}/_clear_cache`. This is not really a user-facing change and should be fine in a minor.

Reply:
Yup, I understand the fix for this -- my point is that it seems like an arbitrary limitation. Routes with the same prefix but different suffix could realistically co-exist with different wildcards, and don't need to be coupled. If this is known and accepted behavior it's all good though! Not a big limitation.

Reply:
> If this is known and accepted behavior

It's definitely know. Maybe largely accepted as well. But probably not set in stone. core-infra will be able to give out better advices.